### PR TITLE
Add a button to allow mediators to mark a request as completed/done…

### LINKED ISSUE #539: Allow mediator to mark a request as Done
Mediator wants to be able to move a request out of "Pending" state without approving it. Applies to all mediating libraries.
- [x] Add a "mark as done" button in the open request, above the table.

mediate - mark done and comment

Mediator clicks "mark as done":
- [x]  button becomes disabled, request is marked as "Done". 
- [x]  ideally, requests that are artificially closed can be differentiated from done-by-approval requests
- [x]  any approve buttons in the request should remain active: mediator should be able to approve an request at any time (see also #512 )

mediate - marked done disabled

"Marked done" vs "actually done", in context:

mediate - marked done

### PR DESCRIPTION
**Pull request by jkeck**
…without approving all items.

Closes #539 

_Marking this as a **[WIP]** because I notice there is something in #514 that may affect the approach we take here_

mark-as-done

### CODE REVIEW

File: app/assets/javascripts/markAsComplete.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/* global alert */
+
+var markAsComplete = (function() {
```
Comment (jkeck) [inline app/assets/javascripts/markAsComplete.js]: It is used in another file, but hound does not seem to catch that.

**Comment by jkeck** [comment]
@jvine a question that came up when I was going over this with @ndushay: Is it okay that there is no way to undo the "Mark as done" action?

**Comment by jvine** [comment]
@jkeck, question back to you:  When a request is marked as done, is it possible to go back and actually change it? like, approve another item, make a comment?

**Comment by jkeck** [comment]
Just as a heads up, we'll want to do a little data migration after we deploy a release with this change.  It should be something along the lines of:

```ruby
MediatedPage.archived.each do |request|
  if request.all_approved?
    request.approval_status = :approved
  else
    request.approval_status = :marked_as_done
  end
  request.save!
end
```

We'll also probably want to do this after work hours (once the folks mediating requests have finished for the day).

**Comment by jkeck** [comment]
And I answered @jvine's question in person, but just for posterity, marking as done does not restrict the request from taking any other action (it's simply a state that is stored in the database).

File: app/assets/javascripts/markAsComplete.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      this.holdingsTable = $(holdingsTable);
+
+      this.onAjaxSuccess();
+      this.onAjaxError();
+    },
+
+    onAjaxSuccess: function() {
+      var _this = this;
+      _this.markAsCompleteForm().on('ajax:success', function(_, request) {
+        if(request.approval_status === 'marked_as_done') {
```
Comment (jkeck) [inline app/assets/javascripts/markAsComplete.js]: This is coming from an active record attribute.

File: app/assets/javascripts/item_approval.js [hunk 3]
```diff
@@ -71,10 +71,14 @@ var itemApproval = (function() {
...
-            .find('[data-behavior="all-approved-note"]')
-            .show();
+        var tableRow =  item.closest('tr.holdings')
+                            .prev('tr');
+
+            tableRow.find('[data-behavior="all-approved-note"]')
+                    .show();
+
+            tableRow.find('[data-behavior="mixed-approved-note"]')
+                    .hide();
```
Comment (ndushay) [inline app/assets/javascripts/item_approval.js]: if I'm reading the above correctly, the indentation is misleading.  Lines starting `tableRow` are at the same level of nesting as `var tableRow`, correct?    There's also no particular reason to put the last `.blah()` part on the next line, is there?   Lastly, there's an extra space after `=` in line 74

File: db/schema.rb [hunk 4]
```diff
@@ -35,8 +35,8 @@
     t.string   "type"
     t.date     "needed_date"
     t.string   "status"
-    t.datetime "created_at",         null: false
-    t.datetime "updated_at",         null: false
+    t.datetime "created_at",                     null: false
+    t.datetime "updated_at",                     null: false
```
Comment (ndushay) [inline db/schema.rb]: @another wacko alignment thing - doesn't really matter here, and probably came from automagic rails stuff anyway.

**Comment by ndushay** [comment]
Please pardon my ignorance, but is there any way the data migration could be part of the db migration in the rails app??   (i'm guessing not or you would have done it that way already)

File: db/schema.rb (hunk 4 above)
Comment (jkeck) [inline db/schema.rb]: Yeah, this is auto-generated by rails.  I figure it is best not to hand-edit the schema.

**Comment by jkeck** [comment]
@ndushay wrt putting data migrations in with rails database migrations, I believe that is typically considered [bad practice](https://robots.thoughtbot.com/data-migrations-in-rails).

I would be happy to add something akin to what's in that blog post though.

**Comment by ndushay** [comment]
ah - thanks for the reference.  I think a temporary rake task is a good solution - should one be added to this PR?

**Comment by jkeck** [comment]
@ndushay: 538dfb611ef0b3a51505393d920cda5eef5b96f9

**Pull request merged**